contracts/SpokePoolV3Periphery.sol

```diff
@@ -11,28 +11,85 @@ import { V3SpokePoolInterface } from "./interfaces/V3SpokePoolInterface.sol";
 import { IERC20Auth } from "./external/interfaces/IERC20Auth.sol";
 import { WETH9Interface } from "./external/interfaces/WETH9Interface.sol";
 
+/**
+ * @title SpokePoolProxy
+ * @notice User should only call the SpokePoolV3Periphery contract functions that require an ERC20.approval through this
+ * contract. This is purposefully a simple passthrough contract so that the user only approves this contract
+ * to pull its assets while the SpokePoolV3Periphery contract can be used to call any calldata on any exchange
+ * that the user wants to. By separating the contract that gets approved from the contract that executes arbitrary
+ * calldata, the SpokePoolPeriphery does not need to validate the calldata that gets executed.
+ * @dev If this proxy didn't exist and users instead approved and interacted directly with the SpokePoolV3Periphery
+ * then users would run the unneccessary risk that another user could instruct the Periphery contract to steal
+ * any approved tokens that the user had left outstanding.
+ */
+contract SpokePoolPeripheryApproveProxy is Lockable {
+    using SafeERC20 for IERC20;
+
+    error LeftoverInputTokens();
+
+    // The SpokePoolPeriphery should be deterministically deployed at the same address across all networks,
+    // so this contract should also be able to be deterministically deployed at the same address across all networks
+    // since the periphery address is the only constructor argument.
+    SpokePoolV3Periphery public immutable SPOKE_POOL_PERIPHERY;
+
+    constructor(SpokePoolV3Periphery _spokePoolPeriphery) {
+        SPOKE_POOL_PERIPHERY = _spokePoolPeriphery;
+    }
+
+    /**
+     * @notice Swaps tokens on this chain via specified router before submitting Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev If swapToken or acrossInputToken are the native token for this chain then this function might fail.
+     * the assumption is that this function will handle only ERC20 tokens.
+     * @param swapToken Address of the token that will be swapped for acrossInputToken.
+     * @param acrossInputToken Address of the token that will be bridged via Across as the inputToken.
+     * @param exchange Address of the exchange contract to call.
+     * @param routerCalldata ABI encoded function data to call on router. Should form a swap of swapToken for
+     * enough of acrossInputToken, otherwise this function will revert.
+     * @param swapTokenAmount Amount of swapToken to swap for a minimum amount of depositData.inputToken.
+     * @param minExpectedInputTokenAmount Minimum amount of received depositData.inputToken that we'll submit bridge
+     * deposit with.
+     * @param depositData Specifies the Across deposit params we'll send after the swap.
+     */
+    function swapAndBridge(
```

Comment:
Question for others:
- Do we need to force other calls on SpokePoolPeriphery to be sent through this proxy because they can be similarly "approval abused"? For example, can `permit`, `receiveWithAuthorization`, and `permit2` be similarly abused? I'm guessing no because these systems have nonces that prevent re-broadcasting signed permits/auths but I'm not sure

Reply:
@mrice32 @bmzig @pxrl

Reply:
`receiveWithAuthorization` should be OK since it enforces that the `msg.sender` to that call is the receiver.

I do think you are right in that we may need to do something clever for permit/permit2. At least for permit, now if the signature is front-run, the actual call to the periphery may also be front-run, and the user will lose funds. So we may just need to rethink how the proxy works or if we should do a proxy. I think something similar would happen with permit2.

Reply:
why would user lose funds? Doesn't the permit include who the `recipient` is and it is signed by the `sender`?

Reply:
I was just thinking that if you front-run the permit signature, there will always be a race between calling `swapAndBridge` honestly and calling `swapAndBridge` with the token as the exchange and `transferFrom` as the routerCalldata. I'm still not certain on permit2, but I'd imagine if we can front-run the permit2 signature, then it's basically the same thing.

Reply:
i guess as long as you put the routerData in the permit2 witness data that the user signs, there is no ability to front-run the permit2 submission?